rfc/001-truncated-block-history.md

```diff
@@ -0,0 +1,87 @@
+
+* [State sync](https://github.com/tendermint/tendermint/issues/828): bootstraps a new node by syncing the state machine at a given height, but not historical blocks and associated data.
+
+To maintain the integrity of the chain, the use of these features must be coordinated such that necessary historical blocks will not be made unavailable or lost forever. In particular:
+
+* Some nodes should have complete block histories, for auditability and querying.
+
+* The majority of nodes should retain blocks longer than the Cosmos SDK unbonding period, for light client verification.
+
+* Some nodes must take and serve state sync snapshots with snapshot intervals less than the block retention periods, to allow new nodes to state sync and then replay blocks to catch up.
+
+* Only a minority of nodes can be state synced within the unbonding period, for light client verification and to serve block histories for catch-up.
+
+However, it is unclear if and how we should enforce this. It may not be possible to technically enforce all of these without knowing the state of the entire network, but it may also be unrealistic to expect this to be enforced entirely through social consensus. This is especially unfortunate since the consequences of misconfiguration can be permanent chain-wide data loss.
+
+The main configuration options involved are:
+
+* Unbonding time (SDK): the duration in which validators can be economically punished for misbehavior.
+
+* Block time (Tendermint): the minimum duration between consecutive blocks.
+
+* Block retention (Tendermint): the number of recent blocks (heights) to retain.
+
+* Snapshot interval (SDK): the interval (in heights) between taking state sync snapshots.
+
+* Snapshot retention (SDK): the number of recent state sync snapshots to retain.
+
+## Proposal
+
+* Unbonding time (SDK): already implemented as a genesis parameter `app_state.staking.params.unbonding_time`. Must now satisfy `unbonding_time < 2 * block.retention * block.time_iota_ms` otherwise errors (factor of 2 since `block.time_iota_ms` does not take into account block execution time).
+
+* Block time (Tendermint): already implemented as a genesis parameter `consensus_params.block.time_iota_ms`.
+
+* Block retention (Tendermint): implement as a new genesis parameter `consensus_params.block.retention` (default 0, i.e. retain all), and add a local configuration option `consensus.prune_blocks` (default off) that allows node operators to either enable or disable block pruning on a per-node basis. Must be at minimum 2, since the current and previous blocks are required for progress.
+
+* Snapshot interval (SDK): implement as a new genesis parameter `app_state.snapshots.interval` (default 0, i.e. disabled). Must satisfy `interval < 2 * block.retention` otherwise errors, to allow state synced nodes to catch up via block replay (factor of 2 to account for time spent taking a snapshot).
```

Comment:
Why is this a genesis parameter whereas the snapshot retention isn't? Also, all child keys under `app_state` are particular to modules, so `app_state.snapshots` is a misnomer. Can this be a local configuration just like snapshot retention?

Reply:
Snapshot retention is something that node operators may want to vary (e.g. if someone wants to archive all state snapshots for historical or auditing purposes). Making the interval a genesis parameter ensures all nodes take snapshots at the same heights, such that there is a higher availability of snapshots peers and thus faster state syncs for larger states.

I'm fine with using a local config option instead, that was my initial plan.

Reply:
I've changed snapshot interval to be a local config option.